correios/models/posting.py

```diff
@@ -407,69 +423,101 @@ def validate(cls,
-            cls._validate_dimension("sum of dimensions", width + height + length, MAX_SIZE)
-            return
-
-        # Volume.TYPE_CYLINDER
+            error_list += Package._validate_envelope(width, height, length, diameter)
+        elif package_type == Package.TYPE_BOX:
+            error_list += Package._validate_box(width, height, length, diameter)
+        else:  # Volume.TYPE_CYLINDER
+            error_list += Package._validate_cylinder(width, height, length, diameter)
+
+        if error_list:
+            raise exceptions.InvalidPackageError(error_list)
+
+    @classmethod
+    def _validate_envelope(cls, width, height, length, diameter) -> list:
+        if any([width, height, length, diameter]):
+            msg = "Envelope package does not use width, height, length and diameter"
+            return [errors.PackageError(msg)]
+        return []
+
+    @classmethod
+    def _validate_box(cls, width, height, length, diameter) -> list:
+        error_list = []
+
+        if diameter:
+            msg = "Box package does not use diameter"
+            error = errors.PackageError(msg)
+            error_list.append(error)
+
+        call_args = [
+            ("width", width, MAX_WIDTH),
+            ("height", height, MAX_HEIGHT),
+            ("length", length, MAX_LENGTH),
+            ("sum of dimensions", width + height + length, MAX_SIZE)
+        ]
+        for args in call_args:
+            error = cls._validate_dimension(*args)
+            if error:
+                error_list.append(error)
+
+        return error_list
+
+    @classmethod
+    def _validate_cylinder(cls, width, height, length, diameter) -> list:
+        error_list = []
+
         if width or height:
-            raise exceptions.InvalidPackageDimensionsError(
-                "Cylinder does not use width/height: {}x{}".format(width, height)
-            )
+            msg = "Cylinder package does not use width and height"
+            error = errors.PackageError(msg)
+            error_list.append(error)
+
+        call_args = [
+            ("length", length, MAX_CYLINDER_LENGTH),
+            ("diameter", diameter, MAX_DIAMETER),
+            ("cylinder size", length + 2 * diameter, MAX_CYLINDER_SIZE),
+        ]
+        for args in call_args:
+            error = cls._validate_dimension(*args)
+            if error:
+                error_list.append(error)
 
-        cls._validate_dimension("length", length, MAX_CYLINDER_LENGTH)
-        cls._validate_dimension("diameter", diameter, MAX_DIAMETER)
-        cls._validate_dimension("cylinder size", length + 2 * diameter, MAX_CYLINDER_SIZE)
+        return error_list
 
     @classmethod
-    def _validate_dimension(cls, name, value, maximum):
-        msg = "Invalid {} (range 0~{}): {}".format(name, maximum, value)
-        if value <= 0:
-            raise exceptions.InvalidMinPackageDimensionsError(msg)
+    def _validate_dimension(cls, name, value, maximum) -> Union[errors.PackageError, None]:
```

Comment:
I'd like to propose a different signature for this method:

`def _get_dimension_errors(cls, name, value, minimum, maximum) -> List[errors.PackageError]:`

So we can validate min/max values and add errors for minimum size violations. And by returning a list instead of the inconsistent PackageError/NoneType we keep the API cleaner.